### DISCUSSION TITLE
Counting unique viewers

### DISCUSSION THREAD

**Discussion by gabek** [owncast/owncast#374] [Ideas]
I'd like to find a better solution to unique viewer counting.

Currently we count "viewers" technically as "chat users".  So if you're using the embed player, or a external player you don't get counted.  We need a solution that works across more situations.  But each option has caveats and nothing works in all situations.

Previously we would count a combination of websocket users *and* video download clients, but because only some download clients work it was very confusing and error prone.  Anybody using S3 storage would get counted at first when they downloaded the master playlist, and then get removed as a viewer because they would only be downloading from S3.

### Counting a websocket connection as a user

Pros
* We get updates in real-time.
* We can easily map viewers with chat users.

Cons
* If the embed player is being used there is no websocket connected so the viewer isn't counted.
* If an external video player is being used outside of the web websockets aren't even possible.
* If we add this to the embed player then if you embed chat and the player there would be two websocket connections for a single person and you'd be counted twice unless we added some additional flags (chat vs video socket user).

### Counting video download as a user

Pros
* It works with all players.

Cons
* It doesn't work at all if you're using external storage because you're not downloading the video from the owncast server.
* It's not real-time.

### Creating a new standalone Ping API
Pros
* Can be added to embed player without the websocket.

Cons
* Not real-time.
* It doesn't work on external players.

### Downloads with redirect
This is a bit too complicated, but it's an idea.  Where all video downloads hit the owncast server first, so we know you're still an active viewer, and then we redirect the client to where the actual video segment lives (on S3 or wherever).  I don't even know if video players support redirect headers, and this could add additional delay to downloading each segment.

**Comment by geekgonecrazy** [comment]
This might be naive.  But couldn’t we handle downloads of hls playlist to know how many are streaming?

I don’t know the internals too well yet.  But looking at watch.owncast.online

https://watch.owncast.online/hls/stream.m3u8 References: 0/stream.m3u8 which if I understand right it will basically poll because this file is updated with the parts.

So could count downloads of those.  Of course would probably have to keep short cache or cookie so we know it’s a consistent streamer not a new one.

Unless I’m way off the mark with how hls is behaving?

**Reply by gabek** [reply]
That's exactly how the previous version worked, but as of 0.0.3 we let external storage also handle each variant's playlist.  This is because each client will be downloading it every ~4 seconds, and with enough viewers that adds up.  The point of S3 storage is to offload all of that, so that's why everything but the master playlist is now being handed off.  This is also why with 0.0.2 we didn't have this problem haha.

**Reply by geekgonecrazy** [reply]
Ah!  I knew it was too obvious 😂

I’m guessing couldn’t redirect just the m3u8 file and the parts come from s3.

**Reply by gabek** [reply]
Yeah that's what I was thinking.  I'm not sure if it's possible, as I'm not sure if all video players handle redirects, and I'm not sure if the added hop in the request chain will add noticeable delay.  It's something maybe worth experimenting with.

**Comment by gabek** [comment]
As a heads up, we've since added support to the embed player so they're counted, but any 3rd party external players have the same problem as mentioned as above.

**Comment by geekgonecrazy** [comment]
Ok hear me out.. Maybe i'm crazy.  But from what I understand we can actually have m3u8 files loaded from the main m3u8 list?

Could we inject one there with a path to our owncast direct.  Basically this would act as a polling that someone is "streaming"?


So just inject it every so often in the parts list.  It would ping home and we'd know how many were connected?  Is this crazy?
